From cf9b00c2ec6b1a1c2444f1fc1b206b65d90a6b7f Mon Sep 17 00:00:00 2001
From: Alexey Gladkov <legion@kernel.org>
Date: Sun, 15 Dec 2024 13:49:42 +0100
Subject: [PATCH] setfont: Allow arguments and options to be mixed

The new algorithm has regression and the command is no longer valid:

  setfont my-fancy-font.psf.gz -C /dev/tty1

Fixes: aed486b7 ("setfont: Rewrite arguments parser")
Link: https://github.com/legionus/kbd/issues/122
Signed-off-by: Alexey Gladkov <legion@kernel.org>
---
 src/setfont.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/setfont.c b/src/setfont.c
index 92c2408..18eb638 100644
--- a/src/setfont.c
+++ b/src/setfont.c
@@ -70,9 +70,12 @@ kbd_getopt(int argc, char **argv, const struct kbd_option *opts)
 	if (!optind)
 		optind = 1;
 
-	if (optind >= argc || !argv[optind] || argv[optind][0] != '-')
+	if (optind >= argc || !argv[optind])
 		return -1;
 
+	if (argv[optind][0] != '-')
+		return '?';
+
 	option = argv[optind];
 
 	if (option[1] && option[1] != '-') {
@@ -284,8 +287,14 @@ int main(int argc, char *argv[])
 				usage(EXIT_SUCCESS, opthelp);
 				break;
 			case '?':
-				kbd_warning(0, "invalid option '%s'", argv[optind]);
-				usage(EX_USAGE, opthelp);
+				if (argv[optind][0] == '-') {
+					kbd_warning(0, "invalid option '%s'", argv[optind]);
+					usage(EX_USAGE, opthelp);
+				}
+
+				if (ifilct == MAXIFILES)
+					kbd_error(EX_USAGE, 0, _("Too many input files."));
+				ifiles[ifilct++] = argv[optind++];
 				break;
 		}
 	}
